lib/redis.js

```diff
@@ -32,63 +33,77 @@ module.exports = class Redis {
     ];
   }
 
-  constructor(host, ttl) {
-    this._host = host;
-    this._ttl = ttl || 7200;
+  constructor(redisUrl, redisTTL = 7200, maxRetries = 5) {
+    this.disconnected = true;
+    this.url = redisUrl;
+    this.ttl = redisTTL;
+    this.conn = Redis.buildConn(redisUrl, maxRetries);
+    this.conn.on('connect', () => this.disconnected = false);
+    this.conn.on('error', err => {
+      logger.error('Redis error:', err);
+      this.disconnected = true;
+    });
   }
 
-  hostName() {
-    if (this._host) {
-      return 'redis://' + this._host.replace(/^.*\/\//, '').replace(/:.*$/, '');
+  static buildConn(redisUrl, maxRetries) {
+    const opts = {
+      lazyConnect: true,
+      maxRetriesPerRequest: maxRetries,
+      retryStrategy: times => {
+        if (times <= maxRetries) {
+          return Math.min(times * 50, 2000);
+        } else {
+          this.disconnected = true;
+          return null;
+        }
+      }
+    };
+
+    // optionally run in cluster mode
+    if (redisUrl.match(/^cluster:\/\/.+:[0-9]+$/)) {
+      const parts = redisUrl.replace(/^cluster:\/\//, '').split(':');
+      const config = [{host: parts[0], port: parts[1]}];
+      return new IORedis.Cluster(config, {lazyConnect: true, clusterRetryStrategy: opts.retryStrategy, redisOptions: opts});
     } else {
-      return null;
+      return new IORedis(redisUrl, opts);
     }
   }
 
-  connect() {
-    return new Promise((resolve, reject) => {
-      let hostKey = (this._host || '').match(/^redis:/) ? 'url' : 'host';
-      this._client = redis.createClient({
-        connect_timeout: 5000, // don't wait forever
-        [hostKey]: process.env.REDIS_HOST,
-        retry_strategy: (opts) => {
-          if (opts.error && opts.error.code === 'ECONNREFUSED') {
-            logger.error('Redis client error: ECONNREFUSED');
-            reject(opts.error);
-          }
-          if (opts.attempt < 10) {
-            return 100;
-          } else if (opts.error) {
-            reject(opts.error);
-          } else {
-            reject(new Error(`Failed to connect to redis after ${opts.attempt} attempts`));
-          }
-          return undefined;
-        }
-      });
-      this._client.on('error', err => reject(err));
-      this._client.on('connect', () => resolve(this._client));
-    });
+  get connected() {
+    return !this.disconnected;
   }
 
-  disconnect() {
-    return new Promise((resolve, reject) => {
-      this._client.on('error', err => reject(err));
-      this._client.on('end', () => resolve());
-      this._client.quit();
-    });
+  hostName() {
+    if (this.url && !this.url.match(/^(redis)|(cluster):\/\//)) {
+      return 'redis://' + this.url.replace(/^.*\/\//, '').replace(/:[0-9]+$/, '');
+    } else {
+      return (this.url || '').replace(/:[0-9]+$/, '') || null;
+    }
+  }
+
+  async disconnect() {
+    try {
+      this.disconnected = true;
```

Comment:
OK, so this will try/catch the disconnect here, then return false if the disconnect fails, but assume the client is disconnected.

Reply:
Yep, you got it - and this logic is what (through trial and error) i've found to work in the `dovetail-counts-lambda` repo.  Which already runs against a redis cluster.  At some point, might be good to abstract this all into some sort of `ioredis-but-in-a-lambda` package...